concordium-consensus/src/Concordium/TransactionVerification.hs

```diff
@@ -0,0 +1,312 @@
+    let expiry = ID.validTo credential
+    unless (Types.isTimestampBefore now expiry) $ throwError $ NotOk CredentialDeploymentExpired
+    -- check that the credential deployment is not a duplicate
+    exists <- lift (registrationIdExists (ID.credId accountCreation))
+    when exists $ throwError $ NotOk $ CredentialDeploymentDuplicateAccountRegistrationID (ID.credId accountCreation)
+    let credIpId = ID.ipId accountCreation
+    mIpInfo <- lift (getIdentityProvider credIpId)
+    case mIpInfo of
+      -- check that the identity provider exists
+      Nothing -> throwError $ MaybeOk (CredentialDeploymentInvalidIdentityProvider credIpId)
+      Just ipInfo ->
+        case credential of
+          ID.InitialACWP icdi ->
+            -- check signatures for an initial credential deployment
+            unless (A.verifyInitialAccountCreation ipInfo messageExpiry (S.encode icdi)) $ throwError $ NotOk CredentialDeploymentInvalidSignatures
+          ID.NormalACWP ncdi -> do
+            cryptoParams <- lift getCryptographicParameters
+            let ncdv = ID.cdiValues ncdi
+            mArsInfos <- lift (getAnonymityRevokers (OrdMap.keys (ID.cdvArData ncdv)))
+            case mArsInfos of
+              -- check that the anonymity revokers exist
+              Nothing -> throwError $ MaybeOk CredentialDeploymentInvalidAnonymityRevokers
+              Just arsInfos -> do
+                -- if the credential deployment contained an empty map of 'ChainArData' then the result will be 'Just empty'.
+                when (null arsInfos) $ throwError $ MaybeOk CredentialDeploymentInvalidAnonymityRevokers
+                -- check signatures for a normal credential deployment
+                unless (A.verifyCredential cryptoParams ipInfo arsInfos (S.encode ncdi) (Left messageExpiry)) $ throwError $ NotOk CredentialDeploymentInvalidSignatures
+    return $ Ok CredentialDeploymentSuccess)
+
+-- |Verifies a 'ChainUpdate' transaction.
+-- This function verifies the following:
+-- * Checks that the effective time is no later than the timeout of the chain update.
+-- * Checks that sequence number is ok.
+-- * Checks that the 'ChainUpdate' is correctly signed.
+verifyChainUpdate :: TransactionVerifier pv m => Updates.UpdateInstruction -> m VerificationResult
+verifyChainUpdate ui@Updates.UpdateInstruction{..} =
+  either id id <$> runExceptT (do
+    -- Check that the timeout is no later than the effective time,
+    -- or the update is immediate
+    when (Updates.updateTimeout uiHeader >= Updates.updateEffectiveTime uiHeader && Updates.updateEffectiveTime uiHeader /= 0) $
+      throwError $ NotOk ChainUpdateEffectiveTimeBeforeTimeout
+    -- check that the sequence number not too old.
+    nextSN <- lift $ getNextUpdateSequenceNumber (Updates.updateType (Updates.uiPayload ui))
+    let nonce = Updates.updateSeqNumber (Updates.uiHeader ui)
+    when (nextSN > nonce) $ throwError $ NotOk $ ChainUpdateSequenceNumberTooOld nextSN
```

Comment:
Why do we allow any future sequence numbers, in contrast to what we do for account transactions?